app/src/main/java/org/thoughtcrime/securesms/jobs/PushSendJob.java

```diff
@@ -0,0 +1,539 @@
+    }
+  }
+
+  List<SharedContact> getSharedContactsFor(OutgoingMediaMessage mediaMessage) {
+    List<SharedContact> sharedContacts = new LinkedList<>();
+
+    for (Contact contact : mediaMessage.getSharedContacts()) {
+      SharedContact.Builder builder = ContactModelMapper.localToRemoteBuilder(contact);
+      SharedContact.Avatar  avatar  = null;
+
+      if (contact.getAvatar() != null && contact.getAvatar().getAttachment() != null) {
+        avatar = SharedContact.Avatar.newBuilder().withAttachment(getAttachmentFor(contact.getAvatarAttachment()))
+                                                  .withProfileFlag(contact.getAvatar().isProfile())
+                                                  .build();
+      }
+
+      builder.setAvatar(avatar);
+      sharedContacts.add(builder.build());
+    }
+
+    return sharedContacts;
+  }
+
+  List<Preview> getPreviewsFor(OutgoingMediaMessage mediaMessage) {
+    return Stream.of(mediaMessage.getLinkPreviews()).map(lp -> {
+      SignalServiceAttachment attachment = lp.getThumbnail().isPresent() ? getAttachmentPointerFor(lp.getThumbnail().get()) : null;
+      return new Preview(lp.getUrl(), lp.getTitle(), lp.getDescription(), lp.getDate(), Optional.fromNullable(attachment));
+    }).toList();
+  }
+
+  List<SignalServiceDataMessage.Mention> getMentionsFor(@NonNull List<Mention> mentions) {
+    return Stream.of(mentions)
+                 .map(m -> new SignalServiceDataMessage.Mention(Recipient.resolved(m.getRecipientId()).requireAci(), m.getStart(), m.getLength()))
+                 .toList();
+  }
+
+  protected void rotateSenderCertificateIfNecessary() throws IOException {
+    try {
+      Collection<CertificateType> requiredCertificateTypes = SignalStore.phoneNumberPrivacy()
+                                                                        .getRequiredCertificateTypes();
+
+      Log.i(TAG, "Ensuring we have these certificates " + requiredCertificateTypes);
+
+      for (CertificateType certificateType : requiredCertificateTypes) {
+
+        byte[] certificateBytes = SignalStore.certificateValues()
+                                             .getUnidentifiedAccessCertificate(certificateType);
+
+        if (certificateBytes == null) {
+          throw new InvalidCertificateException(String.format("No certificate %s was present.", certificateType));
+        }
+
+        SenderCertificate certificate = new SenderCertificate(certificateBytes);
+
+        if (System.currentTimeMillis() > (certificate.getExpiration() - CERTIFICATE_EXPIRATION_BUFFER)) {
+          throw new InvalidCertificateException(String.format(Locale.US, "Certificate %s is expired, or close to it. Expires on: %d, currently: %d", certificateType, certificate.getExpiration(), System.currentTimeMillis()));
+        }
+        Log.d(TAG, String.format("Certificate %s is valid", certificateType));
+      }
+
+      Log.d(TAG, "All certificates are valid.");
+    } catch (InvalidCertificateException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)